### PR TITLE
Chore(kafka-setup): Add an ansible script to setup kafka along with monitoring

### PR DESCRIPTION
**Pull request by Adarshkumar14**
Setup kafka-cluster on eks

Issues:
  - https://github.com/chaoscarnival/bootcamps/issues/5

This PR is for:

 -  Kudo-based Kafka cluster which includes Kafka statefulset and zookeeper pods.
 -  It also Includes Prometheus and grafana.

### CODE REVIEW

File: custom/app-setup/kafka/ansible_kafka_cleanup.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+- name: "[CleanUp]"
```
Comment (uditgaurav) [inline custom/app-setup/kafka/ansible_kafka_cleanup.yml:1]:
```suggestion
- name: "[CleanUp] Remove kafka cluster along with monitoring"
```

File: custom/app-setup/kafka/ansible_kafka_cleanup.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  tasks:
+          - name: "[CleanUp]: Removing  the deployments of prometheus"
+            community.kubernetes.k8s:
+                    state: absent
+                    name: prometheus-operator
+                    kind: Deployment
+                    namespace: monitoring
+
+          - name: "[CleanUp: Removing Grafana deployment"
```
Comment (uditgaurav) [inline custom/app-setup/kafka/ansible_kafka_cleanup.yml:14]:
```suggestion
          - name: "[CleanUp]: Removing Grafana deployment"
```
Check for other tasks as well

File: custom/app-setup/kafka/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    imagePullPolicy: Always
+    envFrom:
+        - secretRef:
+
+              name: aws-secret
+    env:
+        - name: MODE
+          value: "setup"
+        - name: PLATFORM
+          value: "eks"
```
Comment (uditgaurav) [inline custom/app-setup/kafka/README.md:76]: Check for other ENVs like kafka version, os, arch.

File: custom/app-setup/kafka/roles/configure-exporter/.travis.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+---
```
Comment (uditgaurav) [inline custom/app-setup/kafka/roles/configure-exporter/.travis.yml:1]: Do we need this file, if not we can remove it?

File: custom/app-setup/kafka/roles/configure-exporter/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+- name: "[Install]: Installing  kafka exporter service monitor"
+  command: kubectl apply -f service-monitors/kafka-exporter-service-monitor.yaml
+            
+- pause:
+                  seconds: 5
+
+- name: "[Info]: Labeling servicemonitor"
+  shell:
+                    cmd: kubectl label servicemonitor.monitoring.coreos.com kafka -n kafka  app=kafka-exporter --overwrite
```
Comment (uditgaurav) [inline custom/app-setup/kafka/roles/configure-exporter/tasks/main.yml:13]: Can we take namespace as an ENV?

File: custom/app-setup/kafka/roles/configure-kafka/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+              state: present
+    
+- name: "[Info]: Creating an empty file"
+  file:
+        path: kubectl-kudo
+        state: touch
+
+- name: "[Info]: Downloading kubectl-kudo"
+  get_url:
+                url: https://github.com/kudobuilder/kudo/releases/download/v{{ VERSION }}/kubectl-kudo_{{ VERSION }}_{{ OS}}_{{ BIT_VERSION }}
```
Comment (uditgaurav) [inline custom/app-setup/kafka/roles/configure-kafka/tasks/main.yml:16]:
```suggestion
                url: https://github.com/kudobuilder/kudo/releases/download/v{{ VERSION }}/kubectl-kudo_{{ VERSION }}_{{ OS}}_{{ ARCH }}
```

File: custom/app-setup/kafka/README.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+ <table>
+    <tr>
+      <th> Variables </th>
+      <th> Description </th>
+      <th> Specify In Pod </th>
+      <th> Notes </th>
+  </tr>
+  <tr>
+    <td> MODE </td>
+    <td> It defines the mode of experiment </td>
```
Comment (uditgaurav) [inline custom/app-setup/kafka/README.md:69]:
```suggestion
    <td> Provide <code>setup</code> to setup kafka cluster and <code>cleanup</code> for uninstallation </td>
```

File: custom/app-setup/kafka/README.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    <td> Required </td>
+    <td> It supports two value <br/>
+         MODE: setup <br/>
+         MODE: cleanup </td>
+  </tr>
+  <tr>
+    <td> PLATFORM </td>
+    <td> It defines the platform of the k8s cluster </td>
```
Comment (uditgaurav) [inline custom/app-setup/kafka/README.md:77]:
```suggestion
    <td> provide the target platform  name. Like <code>eks</code></td>
```

File: custom/app-setup/kafka/README.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    <td> Optional </td>
+    <td> Currently it supports only eks cluster <br/>
+          PLATFORM: eks </td>
+  </tr>
+  <tr>
+    <td> KUDO_VERSION </td>
+    <td> The Kudo version to Install </td>
```
Comment (uditgaurav) [inline custom/app-setup/kafka/README.md:84]:
```suggestion
    <td> Provide the Kudo version for kafka Installation </td>
```

File: custom/app-setup/kafka/README.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    <td> Optional </td>
+    <td> If KUDO_VERSION is not provided ,By-default It will Install the 0.12.0 version of  KUDO </td>
+  </tr>
+ </table>
```
Comment (uditgaurav) [inline custom/app-setup/kafka/README.md:88]: We can also take the namespace where we want to deploy kafka application as input.

File: custom/app-setup/kafka/ansible_kafka_cleanup.yml [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+            community.kubernetes.k8s:
+                     state: absent
+                     name: monitoring
+                     kind: namespace
+                     
+          - name: "[CleanUp]: Removing prometheus-k8s Statefulsets"
+            community.kubernetes.k8s:
+                     state: absent
+                     name: kafka
+                     kind: namespace
```
Comment (uditgaurav) [inline custom/app-setup/kafka/ansible_kafka_cleanup.yml:39]: When we take Kafka namespace as an input variable then we have to pass that name everywhere as variable in place of hardcodding it as `kafka`

File: custom/app-setup/kafka/ansible_kafka_setup.yml [hunk 12]
```diff
@@ -0,0 +1,32 @@
+- name: "[Install]: Configuring Kafka along with exporter and monitoring]"
+  hosts: localhost
+  vars:
+    ansible_python_interpreter: /usr/bin/python3
+
+  tasks:
+          - name: "[Info]: Configure kafka-sts application"
+            import_role:
```
Comment (uditgaurav) [inline custom/app-setup/kafka/ansible_kafka_setup.yml:8]: can we use 3 spaces only like we use in
```
  vars:    
     ansible_python_interpreter: /usr/bin/python3
```

File: custom/app-setup/kafka/roles/configure-exporter/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,54 @@
+---
+- name: "[Install]: Installing chaos-exporter-service-monitor"
+  command: kubectl apply -f service-monitors/chaos-exporter-service-monitor.yaml
+
+- name: "[Install]: Installing  kafka exporter service monitor"
+  command: kubectl apply -f service-monitors/kafka-exporter-service-monitor.yaml
+            
+- pause:
+                  seconds: 5
```
Comment (uditgaurav) [inline custom/app-setup/kafka/roles/configure-exporter/tasks/main.yml:9]: need to check indentation everywhere once pls.

**Review by uditgaurav — CHANGES_REQUESTED** [review]

File: custom/app-setup/kafka/litmus-kafka-deployer/litmus-kafka-deployer-pod.yml [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+kind: Pod
+metadata:
+  name: litmus-kafka-deployer
+  labels:
+    app: litmus-kafka-deployer
+spec:
+  serviceAccountName: litmus-kafka-sa
+  containers:
+  - name: litmus-kafka-deployer-container
+    image: aady12/demotest:latest
```
Comment (uditgaurav) [inline custom/app-setup/kafka/litmus-kafka-deployer/litmus-kafka-deployer-pod.yml:11]:
```suggestion
    image: litmuschaos/ansible-runner:latest
```

File: custom/app-setup/kafka/roles/configure-exporter/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+- pause:
+     seconds: 5
+
+- name: "[Info]: Labeling servicemonitor"
+  shell:
+     cmd: "kubectl label servicemonitor.monitoring.coreos.com kafka -n {{  KAFKA_NS }}  app=kafka-exporter --overwrite"
+  register: monitor_log
+- debug:
+                  var: monitor_log
```
Comment (uditgaurav) [inline custom/app-setup/kafka/roles/configure-exporter/tasks/main.yml:22]:
```suggestion
- debug:
       var: monitor_log
```

File: custom/app-setup/kafka/roles/configure-kafka/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+- name: "[Wait]: Wait till the  obect of zookeeper is not crerated"
+  community.kubernetes.k8s_info:
+     api_version: apps/v1
+     kind: StatefulSet
+     wait: yes
+     name: zookeeper-instance-zookeeper
+     namespace: '{{ KAFKA_NS }}'
+     wait_sleep: 2
+     wait_timeout: 100
```
Comment (uditgaurav) [inline custom/app-setup/kafka/roles/configure-kafka/tasks/main.yml:54]:
```suggestion
     wait_timeout: 180
```

File: custom/app-setup/kafka/roles/configure-monitoring/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,74 @@
+---
+- name: "[Info]: Cloning the git repository of litmus "
+  git:
+     repo: https://github.com/litmuschaos/litmus.git
+     dest: litmus
+     force: true
```
Comment (uditgaurav) [inline custom/app-setup/kafka/roles/configure-monitoring/tasks/main.yml:6]: can we use release branch

File: custom/app-setup/kafka/roles/configure-monitoring/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+  community.kubernetes.k8s:
+     name: monitoring
+     api_version: v1
+     kind: Namespace
+     state: present
+        
+- name: "[Install]: Installing prometheus-operator"
+  command: kubectl apply -f litmus/monitoring/utils/prometheus/prometheus-operator -n monitoring
+- pause:
+     seconds: 5
```
Comment (uditgaurav) [inline custom/app-setup/kafka/roles/configure-monitoring/tasks/main.yml:18]: do we need this?

File: custom/app-setup/kafka/roles/configure-kafka/vars/main.yml [hunk 19]
```diff
@@ -0,0 +1,6 @@
+---
```
Comment (ksatchit) [inline custom/app-setup/kafka/roles/configure-kafka/vars/main.yml:1]: can we take monitoring info/versions also here? We might also want to take litmus tunables (exporter version).

File: custom/app-setup/kafka/ansible_kafka_cleanup.yml [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+          kind: StatefulSet
+          namespace: '{{ MONITORING_NAMESPACE }}'
+ 
+     - name: "[CleanUp]: Removing '{{ MONITORING_NAMESPACE }}' namespace"
+       community.kubernetes.k8s:
+          state: absent
+          name: '{{ MONITORING_NAMESPACE }}'
+          kind: namespace
+                     
+     - name: "[CleanUp]: Removing prometheus-k8s Statefulsets"
```
Comment (uditgaurav) [inline custom/app-setup/kafka/ansible_kafka_cleanup.yml:36]: can we do a status check for termination?

**Review by uditgaurav — APPROVED** [review]

**Review by uditgaurav — APPROVED** [review]

**Comment by uditgaurav** [comment]
We can also use the provisioned dashboard for grafana with the Kafka metrics. We can also add this is next PRs. @Adarshkumar14

**Comment by Adarshkumar14** [comment]
> We can also use the provisioned dashboard for grafana with the Kafka metrics. We can also add this is next PRs. @Adarshkumar14

File: custom/app-setup/kafka/roles/configure-kafka/tasks/main.yml [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+- name: "[Install]: Installing kafka"
+  command: "kubectl kudo install kafka --instance=kafka -p ADD_SERVICE_MONITOR=true -n {{ KAFKA_NAMESPACE }}"
+  register: kafka_output
+  changed_when: false
+  failed_when:  "kafka_output.rc != 0 and kafka_output.stdout_lines[0].find('already installed') == -1"
+- debug:
+     var: kafka_output.stdout_lines
+
+- name: "[Wait]: Wait till the  object of zookeeper is not crerated"
```
Comment (ispeakc0de) [inline custom/app-setup/kafka/roles/configure-kafka/tasks/main.yml:49]:
```suggestion
- name: "[Wait]: Wait till the  object of zookeeper is not created"
```

**Review by ispeakc0de — APPROVED** [review]

**Review by uditgaurav — APPROVED** [review]

**Pull request merged**